pdf/pdfselection.cpp

```diff
@@ -0,0 +1,486 @@
+        }
+        return -n - boxIndex1;
+    }
+}
+
+int PDFSelection::count() const
+{
+    if (d->pageIndexStart < 0 || d->pageIndexStop < 0 ||
+        d->boxIndexStart < 0 || d->boxIndexStop < 0) {
+        return 0;
+    }
+
+    if (!d->canvas || !d->canvas->document())
+        return 0;
+
+    return d->sliceCount(d->pageIndexStart, d->boxIndexStart,
+                         d->pageIndexStop, d->boxIndexStop) + 1;
+}
+
+PDFCanvas* PDFSelection::canvas() const
+{
+    return d->canvas;
+}
+
+void PDFSelection::setCanvas(PDFCanvas *newCanvas)
+{
+    if (newCanvas != d->canvas) {
+        if (d->canvas)
+            d->canvas->disconnect(this);
+        d->canvas = newCanvas;
+
+        connect(d->canvas, &PDFCanvas::pageLayoutChanged, this, &PDFSelection::handle1Changed);
+        connect(d->canvas, &PDFCanvas::pageLayoutChanged, this, &PDFSelection::handle2Changed);
+        connect(d->canvas, &PDFCanvas::pageLayoutChanged, this, &PDFSelection::onLayoutChanged);
+
+        emit canvasChanged();
+    }
+}
+
+float PDFSelection::wiggle() const
+{
+    return d->wiggle;
+}
+
+void PDFSelection::setWiggle(float newValue)
+{
+    if (newValue != d->wiggle) {
+        d->wiggle = newValue;
+        emit wiggleChanged();
+    }
+}
+
+void PDFSelection::Private::textBoxId(const QPointF &point, Position position,
+                                      int *pageIndex, int *boxIndex)
+{
+    *pageIndex = -1;
+    *boxIndex  = -1;
+    if (!canvas || !canvas->document())
+        return;
+    
+    // point is given in canvas coordinates.
+    QPair<int, QRectF> at = canvas->pageAtPoint(point);
+    if (at.first < 0)
+        return;
+    *pageIndex = at.first;
+
+    const PDFDocument::TextList &boxes = canvas->document()->textBoxesAtPage(*pageIndex);
+    switch (position) {
+    case PDFSelection::Private::At: {
+        *boxIndex = -1;
+        qreal squaredDistanceMin = wiggle * wiggle;
+        int i = 0;
+        for (PDFDocument::TextList::const_iterator box = boxes.begin();
+             box != boxes.end(); box++) {
+            qreal squaredDistance =
+                canvas->squaredDistanceFromRect(at.second, box->first, point);
+                
+            if (squaredDistance < squaredDistanceMin) {
+                *boxIndex = i;
+                squaredDistanceMin = squaredDistance;
+            }
+            i += 1;
+        }
+        break;
+    }
+    case PDFSelection::Private::Before: {
+        QPointF reducedCoordPoint {point.x() / at.second.width(),
+                (point.y() - at.second.y()) / at.second.height()};
+        *boxIndex = -1;
+        for (PDFDocument::TextList::const_iterator box = boxes.begin();
+             box != boxes.end(); box++) {
+            /* Stop counting boxes up as soon as a box after reducedCoordPoint is found. */
+            if (box->first.y() > reducedCoordPoint.y() ||
+                (box->first.y() + box->first.height() > reducedCoordPoint.y() &&
```

Comment:
Better not do too much for this PR, but the assumption doesn't necessarily hold for RTL languages.

Tested: ftp://ctan.tug.org/tex-archive/macros/latex/exptl/mem/arabic.pdf 
-> text gets selected as the flow would be LTR. Then again so does Firefox's internal pdf viewer and Evince. Not sure if pdf itself should have properly ordered boxes or should textDirection() be handled somehow. Maybe some TODO here?